torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py

```diff
@@ -0,0 +1,195 @@
+        if base64_state == self._NULL_SENTINEL.encode():
+            return None
+        try:
+            state = codecs.decode(base64_state, "base64")
+        except binascii.Error as exc:
+            raise RendezvousStateError(
+                "The state object is corrupt. See inner exception for details."
+            ) from exc
+
+        return state, base64_state
+
+
+def _create_tcp_store(params: RendezvousParameters) -> TCPStore:
+    host, port = _parse_rendezvous_endpoint(params.endpoint, default_port=29500)
+
+    cfg_is_host = params.get_as_bool("is_host")
+    # If the user has explicitly specified whether our process should host the
+    # the store, respect it.
+    if cfg_is_host is not None:
+        is_host = cfg_is_host
+    # Otherwise try to determine whether we are the host based on our hostname
+    # and IP address.
+    else:
+        is_host = _matches_machine_hostname(host)
+
+    # The timeout
+    read_timeout = cast(int, params.get_as_int("read_timeout", 60))
+    if read_timeout <= 0:
+        raise ValueError("The read timeout must be a positive integer.")
+
+    # In specific cases we attempt to instantiate the store twice. For details
+    # see the explanation in the except clause below.
+    for _ in range(2):
+        try:
+            store = TCPStore(  # type: ignore[call-arg]
+                host, port, is_master=is_host, timeout=timedelta(seconds=read_timeout)
+            )
+
+            if is_host:
+                log.info(
+                    f"Process {os.getpid()} hosts the TCP store for the C10d rendezvous backend."
+                )
+
+            break
+        except (ValueError, RuntimeError) as exc:
+            # If we heuristically inferred the value of is_host as True and our
+            # first attempt to instantiate the TCP store has failed, try it one
+            # more time with is_host set to False. As an edge case there can be
+            # more than one process that is part of the same rendezvous on this
+            # machine and only one of them will eventually host the store.
+
+            if not is_host or cfg_is_host is not None:
+                raise RendezvousConnectionError(
+                    "The connection to the C10d store has failed. See inner exception for details."
+                ) from exc
+
+            is_host = False
+
+    return store
+
+
+def create_backend(params: RendezvousParameters) -> C10dRendezvousBackend:
+    """Creates a new `C10dRendezvousBackend` from the specified parameters.
+
+    Configuration:
+        store_type (str, optional):
+            The type of the C10d store. As of today the only supported type is
+            "tcp" which corresponds to `TCPStore`. Defaults to "tcp".
+        read_timeout (int, optional):
+            The read timeout, in seconds, for store operations. Defaults to 60
+            seconds.
+        is_host (bool, optional):
```

Comment:
Probably better not have a TCPStore specific flag like this. Maybe we can have an argument like `**store_options` which can be used to specific store specific options?

Reply:
All these "parameters" (`store_type`, `read_timeout`, and `is_host`) are in fact configuration options passed as `**kwargs` to `RendezvousParameters` and read from it using its generic `get()` method. See [here](https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/api.py#L160) and [here](https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/api.py#L182).

`RendezvousParameters` is a backend-agnostic type that has typed attributes only for mandatory parameters such as minimum and maximum number of nodes, and the rendezvous endpoint. Any backend specific configuration is passed via `**kwargs`. The `create_handler()` function is then responsible for parsing its own specific arguments from those `kwargs`.

As I mentioned above `create_handler()` is meant to instantiate a backend from command line arguments which are passed indirectly via `RendezvousParameters`. If the API is used programmatically, one should construct the backend using its `__init__`.

Maybe I should make `create_handler()` internal (meaning `_create_handler()`) to make it more clear that it is not intended to be called in a normal API usage.